hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/resolver/MountTableResolver.java

```diff
@@ -381,18 +417,37 @@ public void clear() {
   public PathLocation getDestinationForPath(final String path)
       throws IOException {
     verifyMountTable();
+    PathLocation res;
+    String tmpPath = path;
+    if (isTrashPath(tmpPath)) {
+      tmpPath = subtractBaseTrashPath(tmpPath);
+    }
+    String finalPath = tmpPath;
     readLock.lock();
     try {
       if (this.locationCache == null) {
-        return lookupLocation(path);
+        res = lookupLocation(finalPath);
+      } else {
+        Callable<? extends PathLocation> meh = new Callable<PathLocation>() {
+          @Override
+          public PathLocation call() throws Exception {
+            return lookupLocation(finalPath);
+          }
+        };
+        res = this.locationCache.get(finalPath, meh);
       }
-      Callable<? extends PathLocation> meh = new Callable<PathLocation>() {
-        @Override
-        public PathLocation call() throws Exception {
-          return lookupLocation(path);
+      if (isTrashPath(path)) {
+        List<RemoteLocation> remoteLocations = new ArrayList<>();
+        for (RemoteLocation remoteLocation : res.getDestinations()) {
+          remoteLocations.add(
+              new RemoteLocation(remoteLocation.getNsId(),
```

Comment:
It might be better to create a new constructor where we pass a RemoteLocation and a path?